src/apis/quipuswap-staking/stableswap-calculations.ts

```diff
@@ -0,0 +1,386 @@
+import { BigNumber } from 'bignumber.js';
+
+import { bigIntClamp } from 'src/utils/big-number.utils';
+import { toIntegerSeconds } from 'src/utils/date.utils';
+
+import { ACCUM_PRECISION, A_PRECISION, FEE_DENOMINATOR, PRECISION } from './consts';
+import {
+  BalancingAccum,
+  PrepareParamsAccum,
+  StableswapFeesStorage,
+  StableswapPool,
+  StableswapTokenInfo,
+  TooLowPoolReservesError
+} from './types';
+
+const NUMERIC_METHODS_EPSILON = 1;
+const MAX_TOKENS_COUNT = 4;
+const MAX_ITERATIONS = 100;
+
+const xpMem = (tokensInfo: StableswapTokenInfo[]) => {
+  return tokensInfo.map(({ rateF, reserves }) => rateF.times(reserves).dividedToIntegerBy(PRECISION));
+};
+
+const getA = (
+  /** timestamp in seconds */
+  t0: BigNumber,
+  a0: BigNumber,
+  /** timestamp in seconds */
+  t1: BigNumber,
+  a1: BigNumber
+) => {
+  const nowTimestamp = new BigNumber(toIntegerSeconds(new Date()));
+
+  if (nowTimestamp.lt(t1)) {
+    const tNum = nowTimestamp.minus(t0);
+    const tDen = t1.minus(t0);
+    const diff = a1.minus(a0).abs();
+    const value = diff.times(tNum).dividedToIntegerBy(tDen);
+
+    return a1.gt(a0) ? a0.plus(value) : a0.minus(value).abs();
+  }
+
+  return a1;
+};
+
+const getD = (xp: BigNumber[], ampF: BigNumber) => {
```

Comment:
sorry, can we add some comments or rename this func to more meaningful? 
it's hard to get what this func has done

Reply:
Some comments have been added